tests/difference/core/quint_model/ccv_statemachine.qnt

```diff
@@ -0,0 +1,506 @@
+
+
+    var currentState: ProtocolState
+
+
+    pure def ACTIONS = Set(
+        "VotingPowerChange",
+        "DeliverVscPacket",
+        "DeliverVscMaturedPacket",
+        "EndAndBeginBlockForProvider",
+        "EndAndBeginBlockForConsumer"
+    )
+
+    // a type storing the parameters used in actions.
+    // this is used in the trace to store
+    // the name of the last action, plus the parameters we passed to it.
+    // Note: This type holds ALL parameters that are used in ANY action,
+    // so not all of these fields are relevant to each action.
+    type Action =
+        {
+            kind: str,
+            consumerChain: Chain,
+            timeAdvancement: Time,
+            consumersToStart: Set[Chain],
+            consumersToStop: Set[Chain],
+            validator: Node,
+            newVotingPower: int,
+        }
+    
+
+    var trace: List[Action]
+
+    // a few different values for time advancements.
+    // to keep the number of possible steps small, we only have a few different values.
+    // Roughly, 1s for very small advances (like between blocks),
+    // and then longer values for increasingly severe downtime scenarios.
+    // Note that these can still be combined, so in effect we can get all time advancements by any amount of seconds.
+    pure val timeAdvancements = Set(1 * Second, 1 * Day, 1 * Week, 4 * Week)
+
+    pure def emptyAction =
+            {
+                kind: "",
+                consumerChain: "",
+                timeAdvancement: 0 * Second,
+                consumersToStart: Set(),
+                consumersToStop: Set(),
+                validator: "",
+                newVotingPower: 0,
+            }
+
+    // step allows the most generic nondeterminism, in particular it becomes relatively likely
+    // that over a long enough runtime, all consumers would time out by mismatching their time advancements,
+    // and each endblock has a good chance to stop consumers, ...
+    // step is thus suited to test also unhappy paths.
+    action step = any {
+        all {
+            runningConsumers.size() > 0, // ensure there is a running consumer, otherwise this action does not make sense
+            nondet chain = oneOf(runningConsumers)
+            nondet timeAdvancement = oneOf(timeAdvancements)
+            EndAndBeginBlockForConsumer(chain, timeAdvancement),
+        },
+
+        val consumerStatus = currentState.providerState.consumerStatus
+        nondet consumersToStart = oneOf(nonConsumers.powerset())
+        nondet consumersToStop = oneOf(runningConsumers.powerset())
+        nondet timeAdvancement = oneOf(timeAdvancements)
+        EndAndBeginBlockForProvider(timeAdvancement, consumersToStart, consumersToStop),
+
+        nondet node = oneOf(nodes)
+        // very restricted set of voting powers. exact values are not important,
+        // and this keeps the state space smaller.
+        // 0 for getting a validator out of the validator set, and two non-zero values
```

Comment:
> 0 for getting a validator out of the validator set,
We do not kick the validator out of the set anymore from what I understand. So, would it make sense to remove this comment?

Reply:
Yes, thanks!